src/contribute/portability.md

```diff
@@ -0,0 +1,164 @@
+---
+layout: default
+title: "Portability Framework"
+permalink: /contribute/portability/
+redirect_from: /portability/
+---
+
+# Portability Framework
+
+* TOC
+{:toc}
+
+## Overview
+
+Interoperability between SDKs and runners is a key aspect of Apache
+Beam. So far, however, the reality is that most runners support the
+Java SDK only, because each SDK-runner combination requires non-trivial
+work on both sides. All runners are also currently written in Java,
+which makes support of non-Java SDKs far more expensive. The
+_portability framework_ aims to rectify this situation and provide
+full interoperability across the Beam ecosystem.
+
+The portability framework introduces well-defined, language-neutral
+data structures and protocols between the SDK and runner. This interop
+layer -- called the _portability API_ -- ensures that SDKs and runners
+can work with each other uniformly, reducing the interoperability
+burden for both SDKs and runners to a constant effort.  It notably
+ensures that _new_ SDKs automatically work with existing runners and
+vice versa.  The framework introduces a new runner, the _Universal
+Local Runner (ULR)_, as a practical reference implementation that
+complements the direct runners. Finally, it enables cross-language
+pipelines (sharing I/O or transformations across SDKs) and
+user-customized execution environments ("custom containers").
+
+The portability API consists of a set of smaller contracts that
+isolate SDKs and runners for job submission, management and
+execution. These contracts use protobufs and gRPC for broad language
+support.
+
+ * **Job submission and management**: The _Runner API_ defines a
+   language-neutral pipeline representation with transformations
+   specifying the execution environment as a docker container
+   image. The latter both allows the execution side to set up the
+   right environment as well as opens the door for custom containers
+   and cross-environment pipelines. The _Job API_ allows pipeline
+   execution and configuration to be managed uniformly.
+
+ * **Job execution**: The _Fn API_ defines an execution-time binary
+   contract between the SDK harness and the runner that describes how
```

Comment:
Maybe introducing before what is or what is the goal of the SDK harness will make this paragraph clearer.

Reply:
Good idea. Thanks!